pilot/cmd/pilot-agent/status/ready/probe.go

```diff
@@ -16,62 +16,87 @@ package ready
+	serverState  uint64
+	versionStats util.Stats
 }
 
 // Check executes the probe and returns an error if the probe fails.
 func (p *Probe) Check() error {
+	// Initialize the ServerState to 2 (PRE_INITIALIZING) because 0 indicates LIVE.
+	if p.lastKnownState == nil {
+		p.lastKnownState = &probeState{
+			serverState: 2,
+		}
+	}
 	// First, check that Envoy has received a configuration update from Pilot.
-	if err := p.checkUpdated(); err != nil {
+	if err := p.checkConfigStatus(); err != nil {
 		return err
 	}
 
-	return p.checkServerInfo()
+	return p.checkServerState()
 }
 
-// checkUpdated checks to make sure updates have been received from Pilot
-func (p *Probe) checkUpdated() error {
-	if p.receivedFirstUpdate {
-		return nil
-	}
+// checkConfigStatus checks to make sure initial configs have been received from Pilot.
+func (p *Probe) checkConfigStatus() error {
+	s, err := util.GetVersionStats(p.LocalHostAddr, p.AdminPort)
 
-	s, err := util.GetStats(p.LocalHostAddr, p.AdminPort)
-	if err != nil {
-		return err
+	// If there is timeout, Envoy may be busy processing the config - just return the last known state.
+	if err != nil && !isTimeout(err) {
+		return multierror.Prefix(err, "failed retrieving Envoy stats:")
+	}
+	if err == nil {
+		p.lastKnownState.versionStats.CDSVersion = s.CDSVersion
+		p.lastKnownState.versionStats.LDSVersion = s.LDSVersion
 	}
 
-	CDSUpdated := s.CDSUpdatesSuccess > 0 || s.CDSUpdatesRejection > 0
-	LDSUpdated := s.LDSUpdatesSuccess > 0 || s.LDSUpdatesRejection > 0
-	if CDSUpdated && LDSUpdated {
-		p.receivedFirstUpdate = true
+	if p.lastKnownState.versionStats.CDSVersion > 0 && p.lastKnownState.versionStats.LDSVersion > 0 {
 		return nil
 	}
 
-	return fmt.Errorf("config not received from Pilot (is Pilot running?): %s", s.String())
+	return fmt.Errorf("config not received from Pilot (is Pilot running?) %s", p.lastKnownState.versionStats.String())
 }
 
-// checkServerInfo checks to ensure that Envoy is in the READY state
-func (p *Probe) checkServerInfo() error {
-	info, err := util.GetServerInfo(p.LocalHostAddr, p.AdminPort)
-	if err != nil {
-		return fmt.Errorf("failed to get server info: %v", err)
+// checkServerState checks to ensure that Envoy is in the READY state
+func (p *Probe) checkServerState() error {
+	state, err := util.GetServerState(p.LocalHostAddr, p.AdminPort)
+
+	// If there is timeout, Envoy may be busy processing the config - just return the last known state.
+	if err != nil && !isTimeout(err) {
+		return multierror.Prefix(err, "failed retrieving Envoy server state stat:")
+	}
+
+	if err == nil {
+		p.lastKnownState.serverState = *state
 	}
 
-	if info.GetState() != admin.ServerInfo_LIVE {
-		return fmt.Errorf("server is not live, current state is: %v", info.GetState().String())
+	if admin.ServerInfo_State(p.lastKnownState.serverState) != admin.ServerInfo_LIVE {
+		return fmt.Errorf("server is not live, current state is: %s", admin.ServerInfo_State(*state).String())
```

Comment:
can the function just not return a pointer so we don't have to dereference this?

Reply:
I was thinking about it - but that needs changes how the stats are parsed (parseStats & processLines functions. So thought it is not worth refactoring those functions. WDYT?

Reply:
Also for returning non pointer, to handle error case I returned pointer - otherwise we have to return some value other than 0 - because 0 is a valid value.